autoload/LSP.vim

```diff
@@ -52,8 +52,45 @@ function! LSP#viewport() abort
 endfunction
 
 function! LSP#position() abort
+  return s:get_position()
+endfunction
+
+function! LSP#range(mode) abort
+  if a:mode ==# 'v'
+    return s:get_visual_selection()
+  endif
+
+  return {
+    \ 'start': s:get_position(),
+    \ 'end': s:get_position(),
+    \ }
+endfunction
+
+function! s:get_position() abort
 	return {
 		\ 'line': LSP#line(),
 		\ 'character': LSP#character(),
 		\ }
 endfunction
+
+function! s:get_visual_selection() abort
+    let [line_start, column_start] = getpos("'<")[1:2]
+    let [line_end, column_end] = getpos("'>")[1:2]
+    let lines = getline(line_start, line_end)
+    if len(lines) == 0
+        return ''
```

Comment:
Does this cause serialization error when later used in serde?

Reply:
I guess it could cause a deserialization error if the user mis-uses this function and ends up with 0 lines there.

I added an error message before the return to signal the user that they are doing something wrong, I can't think of a better way of handling this, as returning a default range is probably misleading, so I guess erroring is the right thing here.